packages/core/src/browser/quick-open/quick-input-service.ts

```diff
@@ -148,28 +148,43 @@ export class QuickInputService {
                 acceptor([new QuickOpenItem({
                     label,
                     run: mode => {
+                        let accepted = false;
                         if (!error && mode === QuickOpenMode.OPEN) {
+                            if (validateInput && lookFor !== undefined) {
+                                Promise.resolve(validateInput(lookFor))
+                                    .then(err => {
+                                        if (err) {
+                                            label = err;
+                                            this.quickOpenService.showDecoration(MessageType.Error);
+                                        } else {
+                                            accepted = true;
+                                        }
+                                    });
+                            } else {
+                                accepted = true;
+                            }
+                        }
+                        if (accepted) {
```

Comment:
This will be always `false` since this is handled in the promise which is not `async`

Reply:
@danarad05

Reply:
@vinokurig 
First, When `accepted == false` - nothing happens in `acceptor` handler.
Second - sequence of actions is as follow:
1. Entering value in quickInput input box - doesn't invoke validation because mode === 0 === PREVIEW and not OPEN.
2. Pressing `Enter` invokes `acceptor` handler with mode == OPEN so that validation will be invoked:
3. `Promise.resolve()` is in fact `splitting` thread to 2 threads - original one continues with `accepted = false` (doing nothing) and second one is when Promise is resolved and then if no error occurred would result in `accepted = true` which would fire `onDidAcceptEmitter`.

So, you are right that first synchronic run through `acceptor` stays `accepted = false` but after Promise is resolved it could be true.  So it is a little tricky but the result is what was requested.

Reply:
@danarad05
> and second one is when Promise is resolved and then if no error occurred would result in accepted = true which would fire onDidAcceptEmitter.

I don't see where the `onDidAcceptEmitter` is fired in the validation promise. The input box can't be accepted: 
![ezgif-3-ea0f7354e68f](https://user-images.githubusercontent.com/7668752/114362429-fcd68080-9b7f-11eb-911f-d1fdbe3d9645.gif)

Reply:
@vinokurig 

First, you are absolutely right - my bad. Thank you for persisting with this issue.

Second, please try these changes as I think they are working as wanted:

**quick-input-service.ts:**

```  import { QuickOpenHideReason } from '../../common/quick-open-service';```
```
acceptor([new QuickOpenItem({
                    label,
                    run: mode => {
                        if (!error && mode === QuickOpenMode.OPEN) {
                            if (validateInput && lookFor !== undefined) {
                                Promise.resolve(validateInput(lookFor))
                                    .then(err => {
                                        if (err) {
                                            label = err;
                                            this.quickOpenService.showDecoration(MessageType.Error);
                                        } else {
                                            this.onDidAcceptEmitter.fire(undefined);
                                            resolve(currentText);
                                            this.quickOpenService.hide(QuickOpenHideReason.ELEMENT_SELECTED);
                                        }
                                    });
                            } else {
                                this.onDidAcceptEmitter.fire(undefined);
                                resolve(currentText);
                                return true;
                            }
                        }
                        return false;
                    }
                })]);
```

**monaco-quick-open-service.ts:**
This is a fix of an existing bug where in runtime `monaco.quickOpen.HideReason` is undefined because enums are not downloaded from monaco (monaco-loader.ts). You can try without this change and see bug occur. 

```
hide(reason?: QuickOpenHideReason): void {
        this.widget.hide(reason as number);
}
```

Reply:
@danarad05 Thank you for the quick response, I confirm that the provided changes work. @ronabet Please apply the changes in your PR.